### ISSUE TITLE
Linking errors for miniz-sys

### ISSUE THREAD

**Issue by kroltan** [alexcrichton/flate2-rs#134]
Environment
====

 - Windows 10
 - MSYS2 -mingw64
 - rustc 1.21.0 (3b72af97e 2017-10-09)
   - Also happens with rustc 1.23.0-nightly (5041b3bb3 2017-11-19)
 - GCC 5.3.0
 - ld 2.25.1

How to reproduce
====
Just use the default structure created by [Amethyst](https://www.amethyst.rs/)'s CLI.

Then, run `cargo run`.

Error
====
```
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\crt2.o" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\rayleigh-a39a57622d98f812.0.o" "-o" "E:\\Projects\\rayleigh\\target\\debug\\deps\\rayleigh-a39a57622d98f812.exe" "E:\\Projects\\rayleigh\\target\\debug\\deps\\rayleigh-a39a57622d98f812.crate.allocator.o" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "E:\\Projects\\rayleigh\\target\\debug\\deps" "-L" "C:\\Users\\kroltan\\.cargo\\registry\\src\\github.com-1ecc6299db9ec823\\dwmapi-sys-0.1.0/x86_64" "-L" "E:\\Projects\\rayleigh\\target\\debug\\build\\miniz-sys-0c75d8060b18ff4c\\out" "-L" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst-c0903cbd3ea648ba.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_utils-255087175a461c06.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_input-554ac3c3e878b196.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_config-1df1be6b8bded36f.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libron-5d9a301769eb3b1b.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_audio-e38d1a9ad1137990.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\librodio-a15087f5b12b42e1.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libhound-23ddb72fa938a3e0.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libclaxon-5f4669ba89ea052c.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libcpal-648757e1bc9cbb84.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libole32-6e8c016bf8c4a1cc.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\liblewton-d2d258c393ef0519.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libogg-217e9c9de44c6c90.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\liblazy_static-8b3fe923edb1f868.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_gltf-6a73d66a28f239a8.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libitertools-626b0515fec9e766.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgltf_utils-dc51638f9f7db19b.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgltf-f5014ed18b7788b5.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgltf_json-31549f566da07641.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libserde_json-ab58f4716b60307f.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libdtoa-94a1ebd98ddacb67.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libitoa-8cca89574ec9180d.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libcgmath-f813703d85dfa4b5.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libbase64-e04a4508465922a2.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libsafemem-1e93ff8f9f597eb4.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libbyteorder-64ef402dbc6e77fd.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_renderer-501320c133760717.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgfx_window_glutin-54b735eb35523a00.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libglutin-bf04efce6e7b62e9.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgfx_device_gl-a6e9673c8d7675c6.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgfx_gl-79cfce33512a5ecc.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libwinit-947a01ea891eaf87.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libdwmapi-d1988482985c655a.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libuser32-3876225d7880b5b6.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgdi32-fc8db0f422919ac0.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libshell32-88fad4a90728da97.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libshared_library-618da4f19412b845.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libwavefront_obj-462881964d569bcd.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libshrev-90dcca3b396e225d.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libimagefmt-601fd4a94a1c0da8.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libflate2-3572d941ecdf6547.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libminiz_sys-6faebd2377684510.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libhetseq-9751fc2bdbae6d4c.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgfx-8c4ec75ddf95afbd.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libgfx_core-a0fb821f011636bf.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libdraw_state-dbcdaba33057d7fa.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libbitflags-658b32ed3af5c7e9.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\liblog-12b4524282c4d815.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_core-beb803c8e6f6360c.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libcgmath-0afdb4ddda7b6012.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libnum_traits-ef39bc545355a670.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libapprox-ed99ab5645bb68c8.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libamethyst_assets-ca2c8db555659c99.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libspecs-4fddd73f02623128.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libfutures-9320d8de9176d8b1.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libtuple_utils-e0c9f3523337d5f2.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libshred-f7bc7fd9d97eb590.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libpulse-128427add3417e12.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libtime-9385b03bb6aa977c.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libarrayvec-5b2c37464d2a7de2.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libnodrop-f0b1c7d5c61a3dea.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libodds-99f0c3c028e6cf5e.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libmopa-c72c50c3deab980f.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libparking_lot-e3a72045cee8bf91.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libparking_lot_core-1a1e2b1a81dcf713.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libkernel32-a620d0d9b2cef0d9.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libwinapi-1453bcf937b5ec52.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libsmallvec-5e5817e2b85dac8b.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libserde-a9cdb74530965e70.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libowning_ref-399c2ad6f7cf0b4f.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libstable_deref_trait-f97ed402cdf510a9.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libhibitset-3bc27062d3c14152.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\librayon-b957e29ec5e779a5.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\librayon_core-fe2da3ee2fe27594.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\librand-2b50357999776efc.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libnum_cpus-de67863e81819392.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\liblibc-4dc7e5a1e3da6de7.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\liblazy_static-96aec21c997d98af.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libcoco-15b19c013c7c80e1.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libscopeguard-58657702b1d3e3b6.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libeither-711cedd22aa48e91.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libatom-390640b57273935e.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libfnv-5ac4357327a27c13.rlib" "E:\\Projects\\rayleigh\\target\\debug\\deps\\libcrossbeam-273b8c3d597b041d.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-032e1c452e2de4d1.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-f4544e7032568897.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-8115485dc2a7ae72.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-863fb1f3088fad80.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc_system-af7b1e134ab5ede1.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-4af742a168c01de9.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd_unicode-e2b4d8b13c8cf98f.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librand-8432a2d911f96a2f.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-6fe1fb0e133236a4.rlib" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-297864ab139f9899.rlib" "-Wl,-Bdynamic" "-l" "ole32" "-l" "opengl32" "-l" "dwmapi" "-l" "user32" "-l" "gdi32" "-l" "shell32" "-l" "kernel32" "-l" "advapi32" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,-Bstatic" "-l" "gcc_eh" "-l" "pthread" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\Users\\kroltan\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: E:\Projects\rayleigh\target\debug\deps\libminiz_sys-6faebd2377684510.rlib(miniz.o): In function `tinfl_decompress':
          /c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:1707: undefined reference to `__assert_func'
          E:\Projects\rayleigh\target\debug\deps\libminiz_sys-6faebd2377684510.rlib(miniz.o): In function `tdefl_start_dynamic_block':
          /c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:2024: undefined reference to `__assert_func'
          /c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:2026: undefined reference to `__assert_func'
          /c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:2027: undefined reference to `__assert_func'
          /c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:2030: undefined reference to `__assert_func'
          E:\Projects\rayleigh\target\debug\deps\libminiz_sys-6faebd2377684510.rlib(miniz.o):/c/Users/kroltan/.cargo/registry/src/github.com-1ecc6299db9ec823/miniz-sys-0.1.10/miniz.c:2031: more undefined references to `__assert_func' follow
```

**Comment by alexcrichton** [comment]
This is coming up due to a mismatch in mingw runtime environments, the compiler comes with a bundled mingw linker and if that's removed it should fix the error.

**Comment by kroltan** [comment]
Thanks for the response. With the help of some fine folks over IRC, I've found the location of the bundled linker. But if I remove it, then other crates break, with the following error:

    error: A dynamic link library (DLL) initialization routine failed. (os error 1114)

Is there an alternative fix?

**Comment by alexcrichton** [comment]
I personally "cross compile" where the host compiler is MSVC and the target is MinGW

**Comment by oyvindln** [comment]
With the just published 1.0, you could get around it by using the new rust back-end of flate2 if you can override what features flate2 use.

**Comment by kroltan** [comment]
I ended up switching my toolchain over to MSVC and use that directly. If i _do_ need MinGW in the future, I'll do as Mr. Crichton suggested.
